arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,1235 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdint.h>
+#include <string.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/toolchain.h>
+#include <zephyr/arch/xtensa/arch_inlines.h>
+#include <zephyr/arch/xtensa/mpu.h>
+#include <zephyr/linker/linker-defs.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/util_macro.h>
+
+#include <xtensa/corebits.h>
+#include <xtensa/config/core-matmap.h>
+#include <xtensa/config/core-isa.h>
+#include <xtensa_mpu_priv.h>
+
+extern char _heap_end[];
+extern char _heap_start[];
+extern char __data_start[];
+extern char __data_end[];
+
+#ifndef CONFIG_XTENSA_MPU_ONLY_SOC_RANGES
+__weak const struct xtensa_mpu_range xtensa_soc_mpu_ranges[0];
+__weak int xtensa_soc_mpu_ranges_num;
+#endif
+
+/** MPU foreground map for kernel mode. */
+static struct xtensa_mpu_map xtensa_mpu_map_fg_kernel;
+
+/*
+ * Additional information about the MPU maps: foreground and background
+ * maps.
+ *
+ *
+ * Some things to keep in mind:
+ * - Each MPU region is described by TWO entries:
+ *   [entry_a_address, entry_b_address). For contiguous memory regions,
+ *   this should not much of an issue. However, disjoint memory regions
+ *   "waste" another entry to describe the end of those regions.
+ *   We might run out of available entries in the MPU map because of
+ *   this.
+ *   - The last entry is a special case as there is no more "next"
+ *     entry in the map. In this case, the end of memory is
+ *     the implicit boundary. In another word, the last entry
+ *     describes the region between the start address of this entry
+ *     and the end of memory.
+ * - Current implementation has following limitations:
+ *   - All enabled entries are grouped towards the end of the map.
+ *   - No disabled MPU entries allowed in between.
+ *   - Background map entries are copied to foreground map to
+ *     ensure correct alignment for hardware... which means it may
+ *     waste some foreground entries if there are no custom regions
+ *     between background entries.
+ *
+ *
+ * For foreground map to be valid, its entries must follow these rules:
+ *
+ * - The start addresses must always be in non-descending order.
+ * - The access rights and memory type fields must contain valid values.
+ * - The segment field needs to be correct for each entry.
+ * - MBZ fields must contain only zeroes.
+ * - Although the start address occupies 27 bits of the register,
+ *   it does not mean all 27 bits are usable. The macro
+ *   XCHAL_MPU_ALIGN_BITS provided by the toolchain indicates
+ *   that only bits of and left of this value are valid. This
+ *   corresponds to the minimum segment size (MINSEGMENTSIZE)
+ *   definied in the processor configuration.
```

Comment:
This also seems like needless complication?  No hardware exists AFAIK with a MPU page size other than 4k, and IIRC there's language in the databook that implies strongly this is the only possible setup.

Reply:
I am mostly getting the info from the ISA Reference Manual where segment size can be as small as 32 bytes.

Which databook is it? From what I have:
* LX6 one does not have a MPU chapter.
* The NX one mentioned that the configurable address granularity is minimally 4KB.
* On LX7, it depends if MPU Small Page option is enabled on hardware. If not, that it is 4KB. If enabled, then the segment size depends on fetch width, cache line size, etc.